doc: dt: split into multiple pages and add overarching design goals

mbolivar opened the pull request:
Zephyr's devicetree documentation has gotten a bit unwieldy, and is missing a single page which defines the overall goals of the effort.

Fix that by splitting it into multiple pages, doing a bit of cleanup on the result, and adding that single page.

Fixes:  #22557

mbolivar-nordic committed:
doc: split devicetree docs into multiple pages

The one page on devicetree is too long. Split it into multiple pages
to make it easier to digest and more squintable. This is basically
just moving content around; minimal changes have been made apart from
redoing some transitions and adding a couple of introductory paragraphs.

Rename the 'device-tree' Sphinx :ref: target while we are here.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: clean up introduction

Combine various bits of information that were formerly scattered about
into a logical order, and fix a few mistakes.

Make some policy changes, e.g. discouraging the use of fixup macros.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: add design goals defining the scope of DT

Add a page describing the high-level design goals for how Zephyr
should use DT, with examples and counter-examples from current
practice.

Add a TBD section for code generation. It's not clear (to me at least)
where the discussion on that has landed.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
> Fix that by splitting it into multiple pages, doing a bit of cleanup on the result, and adding that single page.

Just my attempt based on meeting minutes and other discussions, of course :).

henrikbrixandersen reviewed (approved):
This is a huge improvement. Thank you.

carlescufi reviewed (approved):
Looks great @mbolivar. Nothing to add here.

MaureenHelm commented on doc/guides/dts/bindings.rst:
```diff
@@ -0,0 +1,139 @@
+###################
+
+A :ref:`devicetree` on its own is only half the story for describing the
+available hardware devices. Thee tree itself doesn't tell the build system
```
s/Thee/The/

mbolivar replied:
Fixed, thank you.

nashif commented on doc/guides/dts/design.rst:
```diff
@@ -0,0 +1,92 @@
+Example remaining work
+======================
+
+Zephyr's :ref:`sanitycheck_script` currently uses :file:`board.yaml` files to
```
nit: imo this list should be bulleted :)

mbolivar replied:
I've made all the example lists bulleted. It looked weird to have some but not others bulleted when playing around with this. Also fixed a typo on this page.

mbolivar-nordic committed:
doc: dts: pacify checkpatch in macros.rst

This is old text, but checkpatch is complaining since it was
introduced into itso wn file.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
I added a patch to satisfy checkpatch, but don't want to do too much more cleanup work in this PR. I will send a separate PR on top of this one with more cleanups.
